File: examples/python/_tls.py
"""Shared TLS auto-detect for the example clients.

Both `client.py` and `phone_call.py` call `make_channel(target)` to open a
gRPC channel that picks up the example self-signed cert at
`examples/certs/server.crt` automatically. Override via the `TLS_CA` env var
(point it at any PEM file). Set it to a path that doesn't exist to force
plain-text.
"""

from __future__ import annotations

import os
from pathlib import Path

import grpc

_DEFAULT_TLS_DIR = Path(__file__).resolve().parent.parent / "certs"


def _load_ca() -> bytes | None:
    ca_path = Path(os.environ.get("TLS_CA", _DEFAULT_TLS_DIR / "server.crt"))
    if ca_path.is_file():
        return ca_path.read_bytes()
    return None


def _load_client_keypair() -> tuple[bytes, bytes] | None:
    """Return (key_bytes, cert_bytes) for the client cert when MTLS=1, else None.

    Defaults to examples/certs/client.{crt,key}. Override via TLS_CLIENT_CERT /
    TLS_CLIENT_KEY. If MTLS=1 but either file is missing, returns None — the
    caller falls back to plain TLS and prints a notice via transport_label().
    """
    if os.environ.get("MTLS", "").lower() not in ("1", "true", "yes"):
        return None
    cert_path = Path(os.environ.get("TLS_CLIENT_CERT", _DEFAULT_TLS_DIR / "client.crt"))
    key_path = Path(os.environ.get("TLS_CLIENT_KEY", _DEFAULT_TLS_DIR / "client.key"))
    if cert_path.is_file() and key_path.is_file():
        return key_path.read_bytes(), cert_path.read_bytes()
    return None


def _channel_credentials() -> grpc.ChannelCredentials | None:
    ca = _load_ca()
    if ca is None:
        return None
    client = _load_client_keypair()
    if client is not None:
        key, cert = client
        return grpc.ssl_channel_credentials(
            root_certificates=ca, private_key=key, certificate_chain=cert,
        )
    return grpc.ssl_channel_credentials(root_certificates=ca)


def make_aio_channel(target: str) -> grpc.aio.Channel:
    """Async channel (grpc.aio)."""
    creds = _channel_credentials()
    if creds is not None:
        return grpc.aio.secure_channel(target, creds)
    return grpc.aio.insecure_channel(target)


def make_sync_channel(target: str) -> grpc.Channel:
    """Sync channel (grpc.Channel) — for `client.py`'s blocking stub."""
    creds = _channel_credentials()
    if creds is not None:
        return grpc.secure_channel(target, creds)
    return grpc.insecure_channel(target)


def transport_label() -> str:
    """Short string for the start-of-run header line."""
    if _load_ca() is None:
        return "insecure (no examples/certs/server.crt found)"
    if _load_client_keypair() is not None:
        return "mTLS (self-signed, examples/certs/)"
    if os.environ.get("MTLS", "").lower() in ("1", "true", "yes"):
        return "TLS (self-signed, examples/certs/) — MTLS=1 but client.{crt,key} missing, falling back"
    return "TLS (self-signed, examples/certs/)"

File: examples/typescript/common/tls.ts
// Shared TLS auto-detect for the example server + clients.
//
// Both sides look at examples/certs/server.{crt,key} (committed to the repo
// so the example is TLS-by-default). Override with TLS_CERT / TLS_KEY /
// TLS_CA env vars, or point them at non-existent paths to force insecure
// mode.

import * as fs from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";

import {
  ServerCredentials,
  credentials,
  type ChannelCredentials,
  type ServerCredentials as ServerCredentialsType,
} from "@grpc/grpc-js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
// __dirname → examples/typescript/common/.  Up twice to examples/, then certs/
// (one more `..` than when this file lived at examples/typescript/tls.ts).
const DEFAULT_TLS_DIR = path.resolve(__dirname, "..", "..", "certs");

function mtlsRequested(): boolean {
  return ["1", "true", "yes"].includes((process.env.MTLS ?? "").toLowerCase());
}

function certPath(): string {
  return process.env.TLS_CERT ?? path.join(DEFAULT_TLS_DIR, "server.crt");
}

function keyPath(): string {
  return process.env.TLS_KEY ?? path.join(DEFAULT_TLS_DIR, "server.key");
}

function caPath(): string {
  return process.env.TLS_CA ?? path.join(DEFAULT_TLS_DIR, "server.crt");
}

function clientCertPath(): string {
  return process.env.TLS_CLIENT_CERT ?? path.join(DEFAULT_TLS_DIR, "client.crt");
}

function clientKeyPath(): string {
  return process.env.TLS_CLIENT_KEY ?? path.join(DEFAULT_TLS_DIR, "client.key");
}

function clientCaPath(): string {
  return process.env.TLS_CLIENT_CA ?? path.join(DEFAULT_TLS_DIR, "client.crt");
}

export function tlsAvailableForServer(): boolean {
  return fs.existsSync(certPath()) && fs.existsSync(keyPath());
}

export function tlsAvailableForClient(): boolean {
  return fs.existsSync(caPath());
}

function mtlsAvailableForServer(): boolean {
  return mtlsRequested() && fs.existsSync(clientCaPath());
}

function mtlsAvailableForClient(): boolean {
  return mtlsRequested() && fs.existsSync(clientCertPath()) && fs.existsSync(clientKeyPath());
}

export function serverCredentials(): ServerCredentialsType {
  if (!tlsAvailableForServer()) return ServerCredentials.createInsecure();
  const requireClientCert = mtlsAvailableForServer();
  return ServerCredentials.createSsl(
    requireClientCert ? fs.readFileSync(clientCaPath()) : null,
    [{ private_key: fs.readFileSync(keyPath()), cert_chain: fs.readFileSync(certPath()) }],
    requireClientCert,
  );
}

export function channelCredentials(): ChannelCredentials {
  if (!tlsAvailableForClient()) return credentials.createInsecure();
  const rootCa = fs.readFileSync(caPath());
  if (mtlsAvailableForClient()) {
    return credentials.createSsl(
      rootCa,
      fs.readFileSync(clientKeyPath()),
      fs.readFileSync(clientCertPath()),
    );
  }
  return credentials.createSsl(rootCa);
}

export function transportLabel(side: "server" | "client"): string {
  if (side === "server") {
    if (!tlsAvailableForServer()) {
      return "insecure (no examples/certs/server.crt found)";
    }
    if (mtlsAvailableForServer()) return "mTLS (self-signed, examples/certs/)";
    if (mtlsRequested()) {
      return "TLS (self-signed, examples/certs/) — MTLS=1 but examples/certs/client.crt missing, falling back to plain TLS";
    }
    return "TLS (self-signed, examples/certs/)";
  }
  if (!tlsAvailableForClient()) {
    return "insecure (no examples/certs/server.crt found)";
  }
  if (mtlsAvailableForClient()) return "mTLS (self-signed, examples/certs/)";
  if (mtlsRequested()) {
    return "TLS (self-signed, examples/certs/) — MTLS=1 but client.{crt,key} missing, falling back";
  }
  return "TLS (self-signed, examples/certs/)";
}
